### PR TITLE
feat(data-warehouse): implement cohere import source

### PR DESCRIPTION
**Pull request by Gilbert09**
## Problem

Cohere was registered as a scaffolded data warehouse source (empty stub) but could not sync anything. Cohere is an enterprise LLM platform, and users who run their AI stack on it want to warehouse their Cohere assets and job history alongside the rest of their data.

## Changes

Implements the `cohere` REST source end-to-end, following the `implementing-warehouse-sources` skill's `source.py` / `settings.py` / `cohere.py` split:

- **Five list endpoints** a user actually wants to warehouse: `datasets`, `connectors`, `models` (catalog), `finetuned_models`, and `embed_jobs`.
- **Full refresh only.** Cohere exposes no reliable server-side timestamp range filter across these endpoints, and the entities are small and mutable (statuses/timestamps change in place), so incremental/append would drop updates. This matches the realistic sync strategy for the API.
- **Per-endpoint pagination** handled declaratively in `settings.py`: offset/limit (`datasets`, `connectors`), `page_size`/`page_token` (`models`, `finetuned_models`), and unpaginated (`embed_jobs`).
- **Partitioning** by a stable `created_at` where present; the model catalog has no creation timestamp so it is left unpartitioned. Primary keys are per-endpoint (`id`, `name` for the model catalog, `job_id` for embed jobs).
- All outbound HTTP goes through `make_tracked_session()`; `tenacity` retries on `429`/`5xx`/transient network errors.
- `validate_credentials`, `get_schemas`, `source_for_pipeline`, and `get_non_retryable_errors` (401/403) implemented. `lists_tables_without_credentials = True` (static catalog) plus `canonical_descriptions.py` so the public docs render the table list.
- Kept behind `unreleasedSource=True` with `releaseStatus=ReleaseStatus.ALPHA`.
- `SOURCES.md` updated (moved `cohere` into the Implemented table).

Endpoint existence, base URL (`https://api.cohere.com/v1`), and auth/error behavior were verified with curl against the live API. Response envelopes and pagination semantics could not be exercised without a paid API key, so parsing is conservative and the uncertainty is noted in code comments.

## How did you test this code?

Added two test modules (36 tests, all passing):

- `tests/test_cohere_source.py` — source-class behavior: source type, config identity/fields, full-refresh schemas, name filtering, public-docs table catalog, non-retryable errors, credential validation, `source_for_pipeline` plumbing, canonical-description coverage.
- `tests/test_cohere.py` — transport: credential status mapping, retryable vs terminal `_fetch_page` behavior, offset pagination short-page termination, `page_token` pagination cursor follow, unpaginated single-request path, and `SourceResponse` shape (primary keys + partitioning) per endpoint.

Ran the targeted tests plus the source-registry-wide `test_source_categories` and `test_source_config_generator` suites (1521 passed). Ran `ruff check --fix` and `ruff format` on the changed files.

Two things could not be done in this environment and need follow-up:

- **Icon:** no `frontend/public/services/cohere.*` asset exists. Fetching one requires a Logo.dev API key, which wasn't available, so no placeholder was committed. `iconPath` still points at `cohere.png`; the asset needs to be added.
- **Docs:** the user-facing doc was written per the `documenting-warehouse-sources` skill (`sourceId: Cohere`, `docsUrl` → `/docs/cdp/sources/cohere`, alpha callout, `<SourceParameters />` + `<SourceTables />`) but lives in the separate posthog.com repo, which wasn't checked out here. It needs to land at `contents/docs/cdp/sources/cohere.md` and pass `audit_source_docs`.
- The `generate_source_configs` management command requires a Postgres connection that wasn't available in this sandbox, so the single-field `CohereSourceConfig` change was applied by hand in the generator's exact output format. The `test_source_config_generator` suite passing confirms it is in sync.

## Automatic notifications

- [ ] Publish to changelog?
- [ ] Alert Sales and Marketing teams?

## Docs update

Doc drafted for posthog.com (see note above); not committed here since posthog.com is a separate repo.

## 🤖 Agent context

**Autonomy:** Human-driven (agent-assisted)

Authored with Claude Code. Skills invoked: `/implementing-warehouse-sources` (end-to-end workflow, architecture split, tracked-transport rule) and `/documenting-warehouse-sources` (doc template).

Key decisions:
- **SimpleSource, full refresh.** Cohere's incremental support is weak (only `/datasets` documents `before`/`after` filters, unverifiable without a paid key) and row counts are tiny, so full refresh is the correct design rather than a client-side cursor that would re-fetch everything anyway.
- **Endpoint set.** Picked the syncable asset/job-history entities (datasets, connectors, model catalog, fine-tuned models, embed jobs) over the mostly-inference API surface. Per-model fine-tuning events (a fan-out) were left out of the alpha to avoid shipping unverifiable fan-out logic; they're a natural follow-up.
- **Pagination modeled as an enum** in `settings.py` so the three families stay declarative and the transport layer branches minimally.
- Verified endpoints/auth with curl; kept envelope parsing conservative where a paid key would be needed to confirm exact shapes.

---
*Created with [PostHog Code](https://posthog.com/code?ref=pr)*

### CODE REVIEW

File: products/warehouse_sources/backend/temporal/data_imports/sources/cohere/cohere.py [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+        "Accept": "application/json",
+    }
+
+
+def validate_credentials(api_key: str) -> bool:
+    # /models is the cheapest authenticated probe: a 200 confirms the key is genuine without
+    # touching a user's data. An invalid key returns 401 ("invalid api token").
+    url = f"{COHERE_BASE_URL}/models"
+    try:
+        response = make_tracked_session().get(url, headers=_get_headers(api_key), params={"page_size": 1}, timeout=10)
```
[the remark this answers is not in the document]
Comment (Gilbert09) [reply products/warehouse_sources/backend/temporal/data_imports/sources/cohere/cohere.py]: Fixed in d9d3145. Both `validate_credentials` and `get_rows` now create their tracked sessions with `redact_values=(api_key,)`, following the `height.py`/`svix.py` pattern, so the bearer token is masked in logged URLs and captured HTTP samples.

File: products/warehouse_sources/backend/temporal/data_imports/sources/cohere/cohere.py [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+
+    return SourceResponse(
+        name=endpoint,
+        items=lambda: get_rows(api_key=api_key, endpoint=endpoint, logger=logger),
+        primary_keys=config.primary_keys,
+        partition_count=1,
+        partition_size=1,
+        partition_mode="datetime" if config.partition_key else None,
+        partition_format="month" if config.partition_key else None,
+        partition_keys=[config.partition_key] if config.partition_key else None,
```
[the remark this answers is not in the document]
Comment (Gilbert09) [reply products/warehouse_sources/backend/temporal/data_imports/sources/cohere/cohere.py]: Fixed in d9d3145. The unpartitioned model catalog now leaves `partition_count`/`partition_size` unset (in addition to mode/format/keys). Previously `partition_count=1` caused the writer to fall back to `primary_keys` and md5-partition the table by `name`; with the count unset it stays unpartitioned as intended. Added assertions to `test_source_response_shape` to guard this.

File: products/warehouse_sources/backend/temporal/data_imports/sources/cohere/cohere.py [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+) -> Iterator[list[dict[str, Any]]]:
+    config = COHERE_ENDPOINTS[endpoint]
+    headers = _get_headers(api_key)
+    # One session reused across every page so urllib3 keeps the connection alive.
+    session = make_tracked_session()
+    url = f"{COHERE_BASE_URL}{config.path}"
+
+    if config.pagination == CoherePagination.NONE:
+        data = _fetch_page(session, url, headers, {}, logger)
+        items = data.get(config.data_key, [])
```
[the remark this answers is not in the document]
Comment (Gilbert09) [reply products/warehouse_sources/backend/temporal/data_imports/sources/cohere/cohere.py]: Fixed in d9d3145. All response parsing now goes through `_extract_items`, which raises `CohereError` when the envelope key is absent rather than yielding an empty page, so a shape mismatch fails loudly instead of clearing the full-refresh table. Covered by `test_missing_envelope_key_raises_instead_of_emptying`.

File: products/warehouse_sources/backend/temporal/data_imports/sources/cohere/cohere.py [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+    pass
+
+
+def _extract_items(data: dict[str, Any], config: CohereEndpointConfig, url: str) -> list[dict[str, Any]]:
+    # A successful response that omits the envelope key is a shape mismatch, not empty data.
+    # Every Cohere schema is full-refresh-only, so silently treating a missing key as an empty
+    # page would clear the existing warehouse table; fail loudly instead of destroying data.
+    if config.data_key not in data:
+        raise CohereError(f"Cohere response missing expected '{config.data_key}' key: url={url}")
+    return data[config.data_key]
```
[the remark this answers is not in the document]
Comment (Gilbert09) [reply products/warehouse_sources/backend/temporal/data_imports/sources/cohere/cohere.py:34]: Thanks, but this is a deliberate design choice rather than a defect, so I am leaving it as-is.

A warehouse import source mirrors the upstream API objects into the customer's *own* data warehouse. The customer owns both the Cohere account and the destination warehouse, so these rows (including any connector OAuth config in their own connectors) are their own data landing in their own store, not data exfiltrated to a third party. Every other source in this directory follows the same convention and yields the API objects unfiltered (see `height.py`, `svix.py`, `stripe.py`) — none maintain a per-endpoint field allowlist.

An allowlist would also be lossy and brittle: it would silently drop fields customers may legitimately want to query, and it would need constant maintenance to keep up as Cohere adds fields (new fields would be dropped by default). The credential-exposure surface we do control — the API key we send on outbound requests — is already masked from logs and captured HTTP samples via `redact_values=(api_key,)`.

Resolving as won't-fix for consistency with the existing source pattern.

**Review by estefaniarabadan — APPROVED** [review]
Isolated new data-warehouse source scaffold — self-contained under its own source directory with only the expected shared-registry additions (SOURCES.md, generated_configs.py, icon), tests included, CI green. Reviewed as part of a batch. LGTM 👍
